tests/test_permissions.py

```diff
@@ -0,0 +1,451 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of Invenio.
+# Copyright (C) 2015-2019 CERN.
+#
+# Invenio is free software; you can redistribute it and/or modify it
+# under the terms of the MIT License; see LICENSE file for more details.
+
+"""Tests for Permission class."""
+
+import pytest
+from flask_principal import ActionNeed, Need, RoleNeed, UserNeed
+from invenio_accounts.models import Role, User
+from invenio_db import db
+
+from invenio_access.models import ActionRoles, ActionSystemRoles, ActionUsers
+from invenio_access.permissions import (
+    ParameterizedActionNeed,
+    Permission,
+    any_user,
+    authenticated_user,
+    superuser_access,
+)
+
+# Test cases summary
+
+# Needs to be tested:
+
+# UserNeed
+# RoleNeed
+# ActionNeed
+# ParameterizedActionNeed
+# SystemRoleNeed
+
+# Test cases:
+
+# 1. allow UserNeed
+# 2. deny UserNeed
+# 3. allow RoleNeed
+# 4. deny RoleNeed
+# 5. allow UserNeed, RoleNeed
+# 6. deny UserNeed, RoleNeed
+# 7. allow ActionNeed -> UserNeed
+# 8. allow ActionNeed -> RoleNeed
+# 9. deny ActionNeed -> UserNeed
+# 10. deny ActionNeed -> RoleNeed
+# 11. allow ActionNeed -> UserNeed, RoleNeed
+# 12. deny ActionNeed -> UserNeed, RoleNeed
+# 13. allow/deny ParameterizedActionNeed
+# 14. test SystemRoleNeed
+# 15. allow_by_default = True/False
+
+
+class WithUserIdentity(object):
+    """Helper class to mock Flask identity and provide needs."""
+
+    def __init__(self, user, *provides):
+        self.user = user
+        self.id = user.id
+        self.provides = set(provides)
+        self.provides.add(UserNeed(user.id))
+        assert ActionNeed not in self.provides
+
+
+def get_superuser():
+    """Create a superuser user."""
+    superuser, = create_users("superpowers")
+    expand(superuser_access, ("allow", superuser))
+    return superuser
+
+
+def create_users(*names):
+    """Helper to create users."""
+    users = []
+    for name in names:
+        user = User(email="{}@inveniosoftware.org".format(name))
```

Comment:
Question: Tests do not send automatic emails right? Just wondering if this might end up in some catch-all mailbox.

Reply:
Ahuhhuhuahuahahua they should not lol! :)